### PR TITLE
`Ibc-rs` documentation

### LINKED ISSUE #1985: `Ibc-rs` and hyperspace documentation
Technical documentation for `ibc-rs` and `hyperspace`.

### PR DESCRIPTION
**Pull request by Wizdave97**
## Issue
Partly closes ComposableFi/centauri#59

### CODE REVIEW

File: code/centauri/ibc/modules/ics02_client.md [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+- Compile the proto files to rust
+- Define the rust equivalents of the structures in the compiled proto files
+- Define conversions between equivalent structs in compiled proto and manually defined
+- Implement `Protobuf` trait for these structs
+- Implement the `ClientState`, `ConsensusState` and `ClientMessage` for the appropriate structs
+- Implement `ClientDef`
+
+
+**Sample Proto file**
+```text
```
Comment (seunlanlege) [inline code/centauri/ibc/modules/ics02_client.md:18]:
```suggestion
```protot
```

File: code/centauri/ibc/modules/README.md [hunk 2]
```diff
@@ -1,47 +1,214 @@
...
+**Mock** contains implementations of the core ibc protocol for testing purposes.  
 
-Implementation of the Inter-Blockchain Communication Protocol ([IBC]) module.
+## Architecture
 
-## Documentation
+The design of this crate is geared towards making it as performant as possible and reducing runtime overhead to the bear minimum.  
+The latter is a major focus because it is expected that this crate will be executed inside blockchain runtimes, which are extremely resource constrained environments  
+and anything that could possibly reduce code execution time needs to be prioritized. 
+To achieve these noble goals, this crate is designed to prioritize static dispatch over dynamic dispatch, while the former is less flexible from a development perspective, it helps could potentially help meet performance
```
Comment (blasrodri) [inline code/centauri/ibc/modules/README.md:18]:
```suggestion
To achieve these noble goals, this crate is designed to prioritize static dispatch over dynamic dispatch, while the former is less flexible from a development perspective, it helps to meet performance 
```

File: code/centauri/ibc/modules/README.md [hunk 3]
```diff
@@ -1,47 +1,214 @@
...
+- **Event** - A struct which when emitted signifies successful processing of a message.
+- **Router** - A type that channels packets to the correct module for handling.
+
+## ICS02 Client Definitions
+
+The client module hosts the trait definitions, messages and handlers for light client implementation.
+
+### On chain Light clients
+
+The Ibc protocol is designed to work on top of light clients, light clients are the foundation on which the protocol is built.
```
Comment (blasrodri) [inline code/centauri/ibc/modules/README.md:40]:
```suggestion
The IBC protocol is designed to work on top of light clients, light clients are the foundation on which the protocol is built.  
```

File: code/centauri/ibc/modules/README.md [hunk 4]
```diff
@@ -1,47 +1,214 @@
...
+## ICS02 Client Definitions
+
+The client module hosts the trait definitions, messages and handlers for light client implementation.
+
+### On chain Light clients
+
+The Ibc protocol is designed to work on top of light clients, light clients are the foundation on which the protocol is built.  
+A light client in simple terms allows a resource constrained environment to follow the consensus of a blockchain,  
+it is able to verify statements concerning the state of the blockchain using information extracted from a block header.  
+For this to be a possibility, the blockchain whose light client is being constructed  is required to have a finality protocol
```
Comment (blasrodri) [inline code/centauri/ibc/modules/README.md:43]:
```suggestion
For this to be a possibility, the blockchain whose light client is being constructed  is required to have a finality protocol.  
```

File: code/centauri/ibc/modules/README.md [hunk 5]
```diff
@@ -1,47 +1,214 @@
...
+The client module hosts the trait definitions, messages and handlers for light client implementation.
+
+### On chain Light clients
+
+The Ibc protocol is designed to work on top of light clients, light clients are the foundation on which the protocol is built.  
+A light client in simple terms allows a resource constrained environment to follow the consensus of a blockchain,  
+it is able to verify statements concerning the state of the blockchain using information extracted from a block header.  
+For this to be a possibility, the blockchain whose light client is being constructed  is required to have a finality protocol  
+(a finality protocol is a means by which a blockchain expresses that state transitions within a block are safe and have a very low probability of been reverted),   
+the light client needs to be continuously updated with a stream of finalized block headers, verifying correctness of the headers and extracting information that can be used to verify state proofs.
```
Comment (blasrodri) [inline code/centauri/ibc/modules/README.md:45]:
```suggestion
The light client needs to be continuously updated with a stream of finalized block headers, verifying correctness of the headers and extracting information that can be used to verify state proofs.
```

File: code/centauri/ibc/modules/ics02_client.md [hunk 6]
```diff
@@ -0,0 +1,365 @@
...
+- Compile the proto files to rust
+- Define the rust equivalents of the structures in the compiled proto files
+- Define conversions between equivalent structs in compiled proto and manually defined
+- Implement `Protobuf` trait for these structs
+- Implement the `ClientState`, `ConsensusState` and `ClientMessage` for the appropriate structs
+- Implement `ClientDef`
+
+
+**Sample Proto file**
+```proto
```
Comment (blasrodri) [inline code/centauri/ibc/modules/ics02_client.md:18]:
```suggestion
```protobuf
```

File: code/centauri/ibc/modules/ics02_client.md [hunk 7]
```diff
@@ -0,0 +1,365 @@
...
+	Grandpa(ics10_grandpa::client_message::ClientMessage),
+	#[ibc(proto_url = "BEEFY_CLIENT_MESSAGE_TYPE_URL")]
+	Beefy(ics11_beefy::client_message::ClientMessage),
+	#[ibc(proto_url = "TENDERMINT_CLIENT_MESSAGE_TYPE_URL")]
+	Tendermint(ics07_tendermint::client_message::ClientMessage),
+}
+
+impl Protobuf<Any> for AnyClientMessage {}
+
+impl TryFrom<Any> for AnyClientMessage {
```
Comment (blasrodri) [inline code/centauri/ibc/modules/ics02_client.md:258]: do we need to show this impl?

File: code/centauri/ibc/modules/ics03_connection.md [hunk 8]
```diff
@@ -0,0 +1,105 @@
+## ICS03_CONNECTION
+
+IBC connections are direct links between chains, it is recommended to have only one open connection between two  
+specific chains, which translates to one connection per light client.
```
Comment (blasrodri) [inline code/centauri/ibc/modules/ics03_connection.md:4]: for cost reasons?

**Review by blasrodri — COMMENTED** [review]
Beautiful, comprehensive and exhaustive docs

File: code/centauri/ibc/modules/README.md [hunk 9]
```diff
@@ -1,47 +1,214 @@
...
 
+**Core** contains the traits and handlers that enable the ibc protocol.  
+**Applications** contains sub protocols that are built off the core ibc.  
+**Mock** contains implementations of the core ibc protocol for testing purposes.  
 
-Implementation of the Inter-Blockchain Communication Protocol ([IBC]) module.
+## Architecture
 
-## Documentation
+The design of this crate is geared towards making it as performant as possible and reducing runtime overhead to the bear minimum.
```
Comment (dzmitry-lahoda) [inline code/centauri/ibc/modules/README.md:15]: > The design of this crate is geared towards making it as performant as possible and reducing runtime overhead to the bear minimum.
This line can be removed :) Sure it was coded like that.

File: code/centauri/ibc/modules/README.md [hunk 10]
```diff
@@ -1,47 +1,214 @@
...
 
-Implementation of the Inter-Blockchain Communication Protocol ([IBC]) module.
+## Architecture
 
-## Documentation
+The design of this crate is geared towards making it as performant as possible and reducing runtime overhead to the bear minimum.  
+The latter is a major focus because it is expected that this crate will be executed inside blockchain runtimes, which are extremely resource constrained environments  
+and anything that could possibly reduce code execution time needs to be prioritized. 
+To achieve these noble goals, this crate is designed to prioritize static dispatch over dynamic dispatch, while the former is less flexible from a development perspective, it helps could potentially help meet performance 
+requirements for blockchain runtimes. We also provide a couple procedural macros in the [`ibc-derive`](/code/centauri/ibc/derive) crate,
```
Comment (dzmitry-lahoda) [inline code/centauri/ibc/modules/README.md:19]: > requirements for blockchain runtimes. We also provide a couple procedural macros in the [`ibc-derive`](/code/centauri/ibc/derive) crate,  
that abstracts away a lot of boilerplate code for the end user.

You may put it into list of crates description

File: code/centauri/ibc/modules/README.md [hunk 11]
```diff
@@ -1,47 +1,214 @@
...
+requirements for blockchain runtimes. We also provide a couple procedural macros in the [`ibc-derive`](/code/centauri/ibc/derive) crate,  
+that abstracts away a lot of boilerplate code for the end user.
+
+The framework is mostly defined as a set of traits that need to be implemented to make use of the message handling capabilities.
+
+## Terminology
+
+A couple definitions to help understand the architecture of this framework
+- **Reader** - A `Reader` is a trait that defines methods that provide read access to the underlying storage of the host.
+- **Keeper** - A `Keeper` trait is one that defines methods that provide write access to the underlying storage of the host.
```
Comment (dzmitry-lahoda) [inline code/centauri/ibc/modules/README.md:28]: Reader <-> Writer . So it is like Reader is reader, but it other side for some reason non conventional Keeper.

File: code/centauri/ibc/modules/README.md [hunk 12]
```diff
@@ -1,47 +1,214 @@
...
+- **Context** - The context is a type that implements all the Reader, Keeper and Routing traits that govern access to the storage of the host and module callbacks.
+- **Handler** - A handler is a function that handles processing of an ibc message type and returns a result.
+- **Event** - A struct which when emitted signifies successful processing of a message.
```
Comment (dzmitry-lahoda) [inline code/centauri/ibc/modules/README.md:31]: Are they no fail events? Only success?

File: code/centauri/ibc/modules/README.md [hunk 13]
```diff
@@ -1,47 +1,214 @@
...
+
+The framework is mostly defined as a set of traits that need to be implemented to make use of the message handling capabilities.
+
+## Terminology
+
+A couple definitions to help understand the architecture of this framework
+- **Reader** - A `Reader` is a trait that defines methods that provide read access to the underlying storage of the host.
+- **Keeper** - A `Keeper` trait is one that defines methods that provide write access to the underlying storage of the host.
+- **Context** - The context is a type that implements all the Reader, Keeper and Routing traits that govern access to the storage of the host and module callbacks.
+- **Handler** - A handler is a function that handles processing of an ibc message type and returns a result.
```
[the remark this answers is not in the document]
Comment (dzmitry-lahoda) [reply code/centauri/ibc/modules/README.md:30]: Routing -> Router

File: code/centauri/ibc/modules/README.md [hunk 14]
```diff
@@ -1,47 +1,214 @@
...
+- **Event** - A struct which when emitted signifies successful processing of a message.
+- **Router** - A type that channels packets to the correct module for handling.
+
+## ICS02 Client Definitions
+
+The client module hosts the trait definitions, messages and handlers for light client implementation.
```
Comment (dzmitry-lahoda) [inline code/centauri/ibc/modules/README.md:36]: hosts trait definitions and handlers clear. so `messages` not clear. how one may host message?

File: code/centauri/ibc/modules/README.md [hunk 15]
```diff
@@ -1,47 +1,214 @@
...
+- [`ClientState`](/code/centauri/ibc/modules/src/core/ics02_client/client_state.rs)
+  - This trait defines all the methods for dealing with the client state for a light client
+- [`ConsensusState`](/code/centauri/ibc/modules/src/core/ics02_client/client_consensus.rs)
+  - This trait defined methods for interacting with the Consensus state
+- [`ClientMessage`](/code/centauri/ibc/modules/src/core/ics02_client/client_message.rs)
+  - This trait defines methods for downcasting to the type contained in the client message enum variants
+
+**The Client Context**
+
+The client context is defined by the [`ClientReader`](/code/centauri/ibc/modules/src/core/ics02_client/context.rs#L24), [`ClientKeeper`](/code/centauri/ibc/modules/src/core/ics02_client/context.rs#L106)
```
Comment (dzmitry-lahoda) [inline code/centauri/ibc/modules/README.md:62]: In start of document trait is just `Reader`, but here it became `ClientReader`

File: code/centauri/ibc/modules/README.md [hunk 16]
```diff
@@ -1,47 +1,214 @@
...
+
+The routing module defines the entry point into the framework through the [`deliver`](/code/centauri/ibc/modules/src/core/ics26_routing/handler.rs#L40) function .
+
+**Routing Context**
+The `Router` trait defines methods that determine how packets are routed to their destination modules in the host
+**ICS26 Context**
+This trait defines how the router is accessed by the Context object
+**Module Callbacks**
+Ibc applications are sub protocols built on top of the core ibc protocol,  
+ibc applications are required to implement the `Module` trait, so they can execute callbacks for processed messages.
```
Comment (dzmitry-lahoda) [inline code/centauri/ibc/modules/README.md:148]: IBC

File: code/centauri/ibc/modules/ics02_client.md [hunk 17]
```diff
@@ -0,0 +1,365 @@
...
+        uint64 weight = 2;
+    }
+
+    // ClientState
+    message ClientState {
+        // Latest chain height
+        uint32 latest_height = 1;
+
+        // current authority set id
+        uint64 current_set_id = 2;
```
Comment (dzmitry-lahoda) [inline code/centauri/ibc/modules/ics02_client.md:38]: current_set_id in context of message ClientState could mean client_current_set_id . So it is not. But documented that current_set_id is current_authority_set_id. Why not too put authority into field name and no need for documentation?

File: code/centauri/ibc/modules/ics02_client.md [hunk 18]
```diff
@@ -0,0 +1,365 @@
...
+
+    impl Protobuf<RawClientMessage> for ClientMessage {} 
+```
+
+**Define the Light Client struct and implement ClientDef**
+
+```rust
+    pub struct TestLightClient;
+    
+    impl ClientDef for LightClient {
```
Comment (dzmitry-lahoda) [inline code/centauri/ibc/modules/ics02_client.md:168]: LightClient -> TestLightClient

File: code/centauri/ibc/modules/ics03_connection.md [hunk 19]
```diff
@@ -0,0 +1,105 @@
...
+IBC connections are direct links between chains, it is recommended to have only one open connection between two  
+specific chains, which translates to one connection per light client.  
+
+### Connection Context
+
+The connection context encapsulates all the storage requirements for connections in the context object.    
+implement the [`ConnectionReader`](/code/centauri/ibc/modules/src/core/ics03_connection/context.rs#L21)  
+and [`ConnectionKeeper`](/code/centauri/ibc/modules/src/core/ics03_connection/context.rs#L51) for the context object
+
+```text
```
Comment (dzmitry-lahoda) [inline code/centauri/ibc/modules/ics03_connection.md:12]: on other documents rust with ... was ``\`rust , not ``\`text

File: code/centauri/ibc/modules/README.md (hunk 9 above)
Comment (KaiserKarel) [reply code/centauri/ibc/modules/README.md:15]: Yeah, I think static dispatch is pretty standard and not considered head-over-heels for performance. If instead we had used dynamic dispatch, unsafe or asm everywhere, an explanation and warning would make sense. But the core::vec docs don't start talking about the noble goal of static dispatch either ;)

File: code/centauri/ibc/modules/ics03_connection.md (hunk 8 above)
Comment (Wizdave97) [reply code/centauri/ibc/modules/ics03_connection.md:4]: No, it's just unnecessary to have multiple connections can be confusing to users.

File: code/centauri/ibc/modules/ics02_client.md (hunk 7 above)
Comment (Wizdave97) [reply code/centauri/ibc/modules/ics02_client.md:258]: I think I'll just summarise it in text format.

File: code/centauri/ibc/modules/README.md (hunk 11 above)
Comment (Wizdave97) [reply code/centauri/ibc/modules/README.md:28]: This naming was adopted `ibc-go`

**Comment by Wizdave97** [comment]
Replaced by https://github.com/ComposableFi/centauri/pull/85

**Pull request closed without merging**